File: frontend/src/api/statusPage.ts
import api from "@/api/scans";

export interface StatusComponent {
  id: string;
  monitor_id: string;
  display_name: string;
  sort_order: number;
  state: string | null;
}

export interface StatusIncidentUpdate {
  id: string;
  body: string;
  status: string;
  created_at: string;
}

export interface StatusIncident {
  id: string;
  title: string;
  impact: string;
  status: string;
  started_at: string;
  resolved_at: string | null;
  created_at: string;
  updates: StatusIncidentUpdate[];
}

export interface StatusPage {
  id: string;
  organization_id: string;
  slug: string;
  title: string;
  published: boolean;
  custom_hostname: string | null;
  hostname_status: string;
  cname_target: string;
  txt_name: string | null;
  txt_value: string | null;
  ssl_status: string | null;
  public_path: string;
  created_at: string;
  updated_at: string;
  components: StatusComponent[];
  incidents: StatusIncident[];
  overall: string | null;
}

export async function getStatusPage(): Promise<StatusPage | null> {
  const { data } = await api.get<StatusPage | null>("/api/status-page");
  return data;
}

export async function upsertStatusPage(payload: {
  slug: string;
  title: string;
}): Promise<StatusPage> {
  const { data } = await api.put<StatusPage>("/api/status-page", payload);
  return data;
}

export async function patchStatusPage(payload: {
  title?: string;
  slug?: string;
  published?: boolean;
  custom_hostname?: string | null;
}): Promise<StatusPage> {
  const { data } = await api.patch<StatusPage>("/api/status-page", payload);
  return data;
}

export async function attachHostname(hostname: string): Promise<StatusPage> {
  const { data } = await api.post<StatusPage>("/api/status-page/hostname", {
    hostname,
  });
  return data;
}

export async function replaceHostname(hostname: string): Promise<StatusPage> {
  const { data } = await api.put<StatusPage>("/api/status-page/hostname", {
    hostname,
  });
  return data;
}

export async function detachHostname(): Promise<StatusPage> {
  const { data } = await api.delete<StatusPage>("/api/status-page/hostname");
  return data;
}

export async function checkHostname(): Promise<StatusPage> {
  const { data } = await api.post<StatusPage>(
    "/api/status-page/hostname/check",
  );
  return data;
}

export async function addComponent(payload: {
  monitor_id: string;
  display_name: string;
  sort_order?: number;
}): Promise<StatusPage> {
  const { data } = await api.post<StatusPage>(
    "/api/status-page/components",
    payload,
  );
  return data;
}

export async function deleteComponent(id: string): Promise<void> {
  await api.delete(`/api/status-page/components/${id}`);
}

export async function createIncident(payload: {
  title: string;
  impact: string;
  status: string;
  body: string;
}): Promise<StatusPage> {
  const { data } = await api.post<StatusPage>(
    "/api/status-page/incidents",
    payload,
  );
  return data;
}

export async function addIncidentUpdate(
  incidentId: string,
  payload: { body: string; status: string },
): Promise<StatusPage> {
  const { data } = await api.post<StatusPage>(
    `/api/status-page/incidents/${incidentId}/updates`,
    payload,
  );
  return data;
}

export async function patchIncident(
  incidentId: string,
  payload: { title?: string; impact?: string },
): Promise<StatusPage> {
  const { data } = await api.patch<StatusPage>(
    `/api/status-page/incidents/${incidentId}`,
    payload,
  );
  return data;
}

export async function deleteIncident(incidentId: string): Promise<void> {
  await api.delete(`/api/status-page/incidents/${incidentId}`);
}

File: backend/app/models/status_page.py
import uuid
from datetime import UTC, datetime
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, CheckConstraint, DateTime, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base

if TYPE_CHECKING:
    from app.models.organization import Organization
    from app.models.uptime import UptimeMonitor
    from app.models.user import User

STATUS_PAGE_PUBLISH_SKUS = frozenset({"pro", "multi"})
STATUS_PAGE_CUSTOM_HOST_SKUS = frozenset({"multi"})
RESERVED_HOST_SUFFIXES = (".sinexis.app",)
PLATFORM_HOSTS = frozenset(
    {
        "sinexis.app",
        "www.sinexis.app",
        "appmedia.id",
        "www.appmedia.id",
        "vs.appmedia.id",
        "www.vs.appmedia.id",
    }
)


class StatusPage(Base):
    __tablename__ = "status_pages"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False
    )
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    slug: Mapped[str] = mapped_column(String(64), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    published: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    custom_hostname: Mapped[str | None] = mapped_column(String(253), nullable=True)
    hostname_status: Mapped[str] = mapped_column(String(24), nullable=False, default="none")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
    )

    organization: Mapped["Organization"] = relationship()
    created_by_user: Mapped["User"] = relationship(foreign_keys=[created_by])
    components: Mapped[list["StatusPageComponent"]] = relationship(back_populates="page", cascade="all, delete-orphan")
    incidents: Mapped[list["StatusIncident"]] = relationship(back_populates="page", cascade="all, delete-orphan")

    __table_args__ = (
        UniqueConstraint("organization_id", name="uq_status_pages_organization_id"),
        UniqueConstraint("slug", name="uq_status_pages_slug"),
        UniqueConstraint("custom_hostname", name="uq_status_pages_custom_hostname"),
        CheckConstraint(
            "hostname_status IN ('none', 'pending_dns', 'active', 'failed')",
            name="ck_status_pages_hostname_status",
        ),
    )


class StatusPageComponent(Base):
    __tablename__ = "status_page_components"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    page_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("status_pages.id", ondelete="CASCADE"), nullable=False, index=True
    )
    monitor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("uptime_monitors.id", ondelete="CASCADE"), nullable=False
    )
    display_name: Mapped[str] = mapped_column(String(255), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    page: Mapped[StatusPage] = relationship(back_populates="components")
    monitor: Mapped["UptimeMonitor"] = relationship()

    __table_args__ = (UniqueConstraint("page_id", "monitor_id", name="uq_status_page_components_page_monitor"),)


class StatusIncident(Base):
    __tablename__ = "status_incidents"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    page_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("status_pages.id", ondelete="CASCADE"), nullable=False, index=True
    )
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    impact: Mapped[str] = mapped_column(String(16), nullable=False, default="minor")
    status: Mapped[str] = mapped_column(String(24), nullable=False, default="investigating")
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))

    page: Mapped[StatusPage] = relationship(back_populates="incidents")
    updates: Mapped[list["StatusIncidentUpdate"]] = relationship(
        back_populates="incident", cascade="all, delete-orphan"
    )

    __table_args__ = (
        CheckConstraint("impact IN ('none', 'minor', 'major', 'critical')", name="ck_status_incidents_impact"),
        CheckConstraint(
            "status IN ('investigating', 'identified', 'monitoring', 'resolved')",
            name="ck_status_incidents_status",
        ),
    )


class StatusIncidentUpdate(Base):
    __tablename__ = "status_incident_updates"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    incident_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("status_incidents.id", ondelete="CASCADE"), nullable=False, index=True
    )
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[str] = mapped_column(String(24), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))

    incident: Mapped[StatusIncident] = relationship(back_populates="updates")
